src/psyclone/psyir/transformations/psy_data_trans.py

```diff
@@ -0,0 +1,192 @@
+        from fparser.two import Fortran2003
+        from fparser.two.utils import walk
+        from psyclone.nemo import NemoInvoke
+        from psyclone.psyGen import OMPDoDirective, ACCLoopDirective
+
+        super(PSyDataTrans, self).validate(node_list, options)
+
+        node_parent = node_list[0].parent
+        if isinstance(node_parent, Schedule) and \
+           isinstance(node_parent.parent, (OMPDoDirective, ACCLoopDirective)):
+            raise TransformationError("A PSyData node cannot be inserted "
+                                      "between an OpenMP/ACC directive and "
+                                      "the loop(s) to which it applies!")
+
+        # The checks below are only for the NEMO API and can be removed
+        # once #435 is done.
+        invoke = node_list[0].root.invoke
+        if not isinstance(invoke, NemoInvoke):
+            return
+
+        # Get the parse tree of the routine containing this region
+        # pylint: disable=protected-access
+        ptree = invoke._ast
+        # pylint: enable=protected-access
+        # Search for the Specification_Part
+        if not walk([ptree], Fortran2003.Specification_Part):
+            raise TransformationError(
+                "For the NEMO API, PSyData can only be added to routines "
+                "which contain existing variable declarations (i.e. a "
+                "Specification Part) but '{0}' does not have any.".format(
+                    invoke.name))
+
+    def apply(self, nodes, options=None):
+        # pylint: disable=arguments-differ
+        '''Apply this transformation to a subset of the nodes within a
+        schedule - i.e. enclose the specified Nodes in the
+        schedule within a single PSyData region.
+
+        :param nodes: can be a single node or a list of nodes.
+        :type nodes: :py:obj:`psyclone.psygen.Node` or list of \
+                     :py:obj:`psyclone.psygen.Node`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :returns: Tuple of the modified schedule and a record of the \
+                  transformation.
+        :rtype: (:py:class:`psyclone.psyGen.Schedule`, \
+                :py:class:`psyclone.undoredo.Memento`)
+
+        '''
+        # Check whether we've been passed a list of nodes or just a
+        # single node.
+        if isinstance(nodes, list) and isinstance(nodes[0], Node):
+            node_list = nodes
+        elif isinstance(nodes, Schedule):
+            # We've been passed a Schedule so default to enclosing its
+            # children.
+            node_list = nodes.children
+        elif isinstance(nodes, Node):
+            # Single node that's not a Schedule
+            node_list = [nodes]
+        else:
+            arg_type = str(type(nodes))
+            raise TransformationError("Error in {1}. "
+                                      "Argument must be a single Node in a "
```

Comment:
Strictly speaking the argument can also be a single Schedule so we should say that too.

Reply:
I have fixed this in the next PR (655-refactor-transformations), since exactly this code was refactored there, which would now cause (imho unnecessary) conflicts.